### ISSUE TITLE
Playing live events with more than one button at the same time (multitimbral keyboard)

### ISSUE THREAD

**Issue by Riobener** [igorski/MWEngine#123]
Apologies for the noob question, your sample project explains how to play live event on a single button. What do I need to do if there are, for example, 3 such buttons, which all pressing at the same time, and it should play a chord, with the sound of the selected synthesizer?
I tried many of my methods, re-read your wiki, and eventually ran into the problem of an excess of memory or the fact that at one time, the synthesizer only plays the last key pressed. 
P.S. sorry for bad english #

**Comment by igorski** [comment]
Hi there, good question!
What you need to do is map your keybaord keys into down/up states (for instance on touch events) and map each unique key to a SynthEvent.
I have [added this page to the Wiki](https://github.com/igorski/MWEngine/wiki/Example-:-creating-a-polyphonic-synthesizer-with-on-screen-keyboard) with some additional information.

**Comment by Riobener** [comment]
Thank you very much! Really appreciate your answer. I will definitely try to do it later.

**Comment by Riobener** [comment]
> Hi there, good question!
> What you need to do is map your keybaord keys into down/up states (for instance on touch events) and map each unique key to a SynthEvent.
> I have [added this page to the Wiki](https://github.com/igorski/MWEngine/wiki/Example-:-creating-a-polyphonic-synthesizer-with-on-screen-keyboard) with some additional information.

Sorry again, i just dont know what am i doing wrong.  I have this code just like in your wiki, in my onCreate method after all permissions:



        // use PWM waveform (see enumeration)
        instrument.getOscillatorProperties( 0 ).setWaveform( 2 );
        ADSR synthEnvelopes = instrument.getAdsr();
        // set a positive release envelopes for a smoother fade out on noteOff
        synthEnvelopes.setReleaseTime( 0.25f );

        for (int i = 0; i < 13; ++i ) {
            int octave = (int) (BASE_OCTAVE + Math.ceil( i / 12 ));
            notes.add( new SynthEvent(( float ) Pitch.note(noteNames.get(i % 12), octave), instrument) );
        }

        pianoView = findViewById(R.id.pianoView);
        pianoView.addPianoTouchListener(new PianoTouchListener() {
            @Override
            public void onKeyDown(@NonNull PianoView piano, int key) {
                // custom to your view, get the key index corresponding to the event
                int index = key;
                // get the synth event
                SynthEvent event = getSynthEventForKeyIndex( index );
                if ( event != null ) {
                    event.play(); // let's heard some sweet sound
                }
            }
            @Override
            public void onKeyUp(@NonNull PianoView piano, int key) {
                // custom to your view, get the key index corresponding to the event
                int index = key;
                // get the synth event
                SynthEvent event = getSynthEventForKeyIndex( index );
                if ( event != null ) {
                    event.stop(); // trigger note end (initializes release phase)
                }
            }

            @Override
            public void onKeyClick(@NonNull PianoView piano, int key) {

            }
        });
And another code above:
```
SynthInstrument instrument = new SynthInstrument();
Vector<SynthEvent> notes = new Vector<SynthEvent>();
int BASE_OCTAVE = 3;
List<String> noteNames = Arrays.asList( "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B" );
```
PianoView is some custom piano i found. It supports click processing normally.

The main issue is the keyboard does not play several jammed notes. Here the result: https://kapwi.ng/c/KhxWd7r7
Thanks in advance!

**Comment by igorski** [comment]
What if you add debug logging to the onKeyDown and onKeyUp handlers ? Visually it looks like the key up and key down should equal the highlighted state of the keys (I'm not familiar with the library). But the audio sounds like the previous note is killed. Or maybe (depending on fidelity during recording), new notes are played but they are all at the same pitch. Can you log the _event.getFrequency()_ just before _event.play()_ ?

**Comment by Riobener** [comment]
Ah, sorry, i accidently delete my previous message. Yeah, i tried. Log is here:

```
2021-03-29 15:30:48.687 28994-28994/com.riobener.audiostudio D/T: event.getFrequency() = 195.9975

2021-03-29 15:30:48.687 28994-28994/com.riobener.audiostudio D/T: KeyNumber = 7

2021-03-29 15:30:48.693 28994-28994/com.riobener.audiostudio D/T: event.getFrequency() = 130.813

2021-03-29 15:30:48.693 28994-28994/com.riobener.audiostudio D/T: KeyNumber = 0

2021-03-29 15:30:48.703 28994-28994/com.riobener.audiostudio D/T: event.getFrequency() = 164.814

2021-03-29 15:30:48.703 28994-28994/com.riobener.audiostudio D/T: KeyNumber = 4
```

I just pressed all three buttons at the same time(Chord C). 

The original code is here: 

```
 // use PWM waveform (see enumeration)
        instrument.getOscillatorProperties( 0 ).setWaveform( 2 );
        ADSR synthEnvelopes = instrument.getAdsr();
        // set a positive release envelopes for a smoother fade out on noteOff
        synthEnvelopes.setReleaseTime( 0.25f );

        for (int i = 0; i < 13; ++i ) {
            int octave = (int) (BASE_OCTAVE + Math.ceil( i / 12 ));
            notes.add( new SynthEvent(( float ) Pitch.note(noteNames.get(i % 12), octave), instrument) );
        }

        pianoView = findViewById(R.id.pianoView);
        pianoView.addPianoTouchListener(new PianoTouchListener() {
            @Override
            public void onKeyDown(@NonNull PianoView piano, int key) {
                // custom to your view, get the key index corresponding to the event
                int index = key;
                // get the synth event
                SynthEvent event = getSynthEventForKeyIndex( index );
                Log.d("T", "event.getFrequency() = "+event.getFrequency());
                Log.d("T", "KeyNumber = "+key);
                if ( event != null ) {
                    event.play(); // let's heard some sweet sound
                }
            }
            @Override
            public void onKeyUp(@NonNull PianoView piano, int key) {
                // custom to your view, get the key index corresponding to the event
                int index = key;
                // get the synth event
                SynthEvent event = getSynthEventForKeyIndex( index );
                if ( event != null ) {
                    event.stop(); // trigger note end (initializes release phase)
                }
            }

            @Override
            public void onKeyClick(@NonNull PianoView piano, int key) {

            }
        });
```

**Comment by Riobener** [comment]
> What if you add debug logging to the onKeyDown and onKeyUp handlers ? Visually it looks like the key up and key down should equal the highlighted state of the keys (I'm not familiar with the library). But the audio sounds like the previous note is killed. Or maybe (depending on fidelity during recording), new notes are played but they are all at the same pitch (there is some audible phase drift). Can you log the _event.getFrequency()_ just before _event.play()_ ?

And it really looks like that the next pressed button is killing sound of previous. It works, if i creating, for example, new unique instrument for each button, but this is crazy solution which always crashes the application cause memory overflow. I'm not sure it is problem with piano keyboard that i found: [https://github.com/convergencelab/PianoView](url). 
All features of your library works perfectly, but polyphony problem is what haunts me so far...

**Comment by igorski** [comment]
Actually a bug has crept in which affected SynthEvents using live playback. This has been addressed in d25df97ae7699db5f836be60677c18c0feeb54ff can you verify after pulling the latest changes whether multi timbral support now works for you ?

**Comment by Riobener** [comment]
> Actually a bug has crept in which affected SynthEvents using live playback. This has been addressed in [d25df97](https://github.com/igorski/MWEngine/commit/d25df97ae7699db5f836be60677c18c0feeb54ff) can you verify after pulling the latest changes whether multi timbral support now works for you ?

Yeah, it finally works for me! I really appreciate your help and a fairly quick response. 
Thank you again :)
